### PR TITLE
Added support for more test methods, including Yarn and Composer

### PR DESCRIPTION
**Pull request by garethr**
This also covers support for methods optionally taking a lock file.

### CODE REVIEW

File: snyk/models.py [hunk 1]
```diff
@@ -176,6 +176,15 @@ def _test(self, path, contents=None):
                 "encoding": "base64",
                 "files": {"target": {"contents": encoded}},
             }
+
+            # Some test methods carry a second file, often a lock file
+            if additional:
```
Comment (nirfuchs) [inline snyk/models.py:181]: Python is [EAFP than LBYL ](https://www.geeksforgeeks.org/eafp-principle-in-python/) therefore a Pythonic way of writing this code block would look like:
```python
try:
    read = getattr(additional, "read", None)
    additional = additional.read()
    encoded = base64.b64encode(additional.encode()).decode()
    post_body["files"]["additional"] = {"contents": encoded}
except:
    pass
```

I'm aware of the fact that the code block at the top of this method is LBYL and therefore this one is too, but worth noting IMHO

File: snyk/models.py [hunk 2]
```diff
@@ -164,7 +164,7 @@ def notification_settings(self):
     def invite(self, email: str, admin: bool = False):
         raise SnykNotImplementedError  # pragma: no cover
 
-    def _test(self, path, contents=None):
+    def _test(self, path, contents=None, additional=None):
```
Comment (nirfuchs) [inline snyk/models.py:167]: How about we add type hints for this method as well, like so:
```python
def _test(self, path: str, contents: typing.IO = None, additional: typing.IO = None)
```

using typing.IO will cover both binary(if necessary) and text file-like objects.

Comment (garethr) [reply snyk/models.py:167]: This won't quite work as is, as the interface accepts file handles or strings.

File: snyk/models.py (hunk 1 above)
Comment (garethr) [reply snyk/models.py:181]: I think I refer the consistency vs having two stanzas with different styles. Not against someone refactoring the whole method mind. It requires some additional type hints I think.

Comment (nirfuchs) [reply snyk/models.py:181]: You have a point there. let's take this one separately as part of a refactor (happy to take it)

File: snyk/models.py (hunk 2 above)
Comment (nirfuchs) [reply snyk/models.py:167]: Thanks for clarifying, API looks ambiguous to me. shouldn't the function calling `_test` be the one reading the file content and passing that on? I guess this is an inherited design smell.

worth a re-factor, though I'm sure we'll have to consider backwards compatibility here.

for instance why `test_pipfile(self, content)` is acting like `test_pipfile_content(self, content) `while what I think would be cleaner is:
```
"""  test_pipfile tests a pip file for issue """
def test_pipfile(self, pip_file: typing.TextIO ) -> bool:
    with open(pip_file) as f:
        self._test(contents=f.read())
```

**Pull request merged**